design/open/repository-operations-error-handling.md

```diff
@@ -0,0 +1,121 @@
+  * A failure in the next creation steps - default `Branch`, initial `Commit` or any other future step, will transfer the `Repository` to state `failed`, if possible. If that cannot be achieved (e.g. the creating process crashed, or KV connectivity is lost) the `Repository` can stay in state `initial` as it is still unused
+  * If during creation, a `Repository` with identical ID (name) is found, the creation of the new `Repository` is stopped before any update the the `KV` is done. This is similar to the current behavior and will be done regardless of the existing `Repository` status
+* When retrieving a `Repository` it will only be retrieved if it is in `active` status. Otherwise an error will be returned.
+  * If the `Repository` entry does not exist, an `ErrRepositoryNotFound` will be returned
+  * If the `Repository` entry exists and is not in `active` state - **TBD #1** - should we return `ErrRepositoryNotFound` here too? Maybe a new error - `ErrRepositoryDeletionInProgress`-ish?
+* Deleting a `Repository` will be done by first setting the `Repository` state to `deleted`, making it immediately unavailable. Once the `Repository` is set to deleted all its correlated entities can be removed safely, as none of them should be accessible. In case of a failure at any step, the `Repository` and all its entities remain unreachable due to the `deleted` status.
+  * A failure to set the `Repository` state to `deleted`, will result in no change to `KV` and so the data remains consistent. The error from the failed update operation will be returned
+  * Once all correlated entities were successfully deleted, the `Repository` entry itself will can be deleted too
+  * Any failure to delete any entity during the deletion process, still leaves the DB in consistent state as the `Repository` entity, and by that its correlated entities, are unreachable due to the `deleted` state
+
+![Repository State Transitions](diagrams/repository-state-transitions.png)
+[(excalidraw file)](diagrams/repository-state-transitions.excalidraw)
+
+### Repository Unique Identifier
+Each `Repository` will have a `unique_identifier` (To be discussed later) which can differ it from other repositories with the same **Name** (`RepositoryID`). This will provide us with the ability to differ object correlated with different repositories with the same **Name**, from different points in time. Each repository will have a designated KV partition, identified by its **Name** and `unique_identifier` and all correlated objects will be handled in this partition, making it possible to identify objects correlated only to an `active` `Repository`. Moreover, as the `unique_identifier` is an attribute of the `Repository`, in order to create the partition name, the `Repository` entity must be acquired, and so non-exist or unusable `Repository` entries will make unreachable partitions and, in turn, unreachable entities, making it safe to use any reachable entity
```

Comment:
It also allows us to have two repositories with the same name, while one is being deleted and the other being created

Reply:
That would be impossible, as (currently) the 2 repositories uses the same key in KV, for the repository entries themselves. That is necessary as the `unique_identifier` is not part of the `Repository` path in KV, as it will not allow direct access.

I had a thought of creating a new key for each deleted/failed repository. Something like `/deleted_repos/<REPO_ID>/<UNIQUE_IDENTIFIER>`, or even a dedicated "deleted" partition, with /<REPOS_ID>_<UNIQ_ID> keys, just as a place holders for cleanup, instead of the /reps/<REPO_ID>.
This will allow multiple instances of the same `RepositoryID`, but only one active under `/repos/<REPOS_ID>`, and will support an immediate "recreation" of a deleted/failed repository, as the /repos/<REPOS_ID> key will be removed
However, that adds another level of complication, as in case of failure, there are 2 more KV actions to perform (addition of a key to the "deleted" partition and removal of the `/repo/<REPO_ID>` key from the common `graveler` partition, and since we got here due to (possibly) a KV failure to start with, this does not solve anything and is a best effort optimization at best